### PR TITLE
[COR-293] return variance on totp comparison

### PR DESCRIPTION
**Pull request by morus12**
Return variance on TOTP comparison.

### CODE REVIEW

**Review by apynes — APPROVED** [review]
Suggestion on naming, otherwise LGTM :+1:

File: crypto/totp.go [hunk 1]
```diff
@@ -98,24 +98,32 @@ func TOTPCompare(key string, period time.Duration, adjust int, length int, chall
...
 // constant time execution is not guaranteed, allows for totp to fall with the variance range of steps + or -
-func TOTPCompareWithVariance(key string, period time.Duration, length int, variance uint, challenge string) (bool, error) {
+func TOTPCompareWithVariance(key string, period time.Duration, length int, variance uint, challenge string) (ok bool, err error) {
+	ok, _, err = TOTPCompareAndGetVariance(key, period, length, variance, challenge)
+	return
+}
+
+// TOTPCompareAndGetVariance the expected TOTP calculation with the challenge in constant time. If variance is > 0
+// constant time execution is not guaranteed, allows for totp to fall with the variance range of steps + or -, variance is returned
+func TOTPCompareAndGetVariance(key string, period time.Duration, length int, variance uint, challenge string) (bool, int, error) {
```
[the remark this answers is not in the document]
Comment (apynes) [reply crypto/totp.go:108]: Didn't notice earlier, this needs to accept the drift as a parameter to center the comparison window on

**Review by apynes — CHANGES_REQUESTED** [review]
Need to accept drift as a parameter

**Review by nmarie22 — APPROVED** [review]

File: crypto/totp.go [hunk 2]
```diff
@@ -98,24 +98,32 @@ func TOTPCompare(key string, period time.Duration, adjust int, length int, chall
...
 // constant time execution is not guaranteed, allows for totp to fall with the variance range of steps + or -
-func TOTPCompareWithVariance(key string, period time.Duration, length int, variance uint, challenge string) (bool, error) {
+func TOTPCompareWithVariance(key string, period time.Duration, length int, variance uint, challenge string) (ok bool, err error) {
+	ok, _, err = TOTPCompareAndGetDrift(key, period, length, variance, challenge)
+	return
+}
+
+// TOTPCompareAndGetDrift the expected TOTP calculation with the challenge in constant time. If variance is > 0
+// constant time execution is not guaranteed, allows for totp to fall with the variance range of steps + or -, variance is returned
+func TOTPCompareAndGetDrift(key string, period time.Duration, length int, variance uint, challenge string) (bool, int, error) {
```
Comment (apynes) [inline crypto/totp.go:108]:
```suggestion
func TOTPCompareAndGetDrift(key string, period time.Duration, length int, variance uint, challenge string, drift int) (bool, int, error) {
```

File: crypto/totp.go [hunk 3]
```diff
@@ -98,24 +98,32 @@ func TOTPCompare(key string, period time.Duration, adjust int, length int, chall
...
 	var eq bool
 	var err error
 	// do 0, +1 && -1, +2 && -2, etc.
 	for vary := 0; vary <= int(variance); vary++ {
```
Comment (apynes) [inline crypto/totp.go:112]: we want to skew this using the drift passed in, so if we have the values we would want to check the frames according to this:
```
drift   variance  frames
0       0              0
0       1              -1,0,1
0       2              -2,-1,0,1,2
1       1              0,1,2
1       2              -1,0,1,2,3
-1      1              -2,-1,0
-1      2              -3,-2,-1,0,1
```

**Review by apynes — CHANGES_REQUESTED** [review]
I think you may have missed my comment that I requested changes about. Added some more clarification. Let me know if this still is not making sense.

File: intutil/math.go [hunk 4]
```diff
@@ -14,48 +15,60 @@ const (
 )
 
 // Abs value of an int
-func Abs(a int) int {
-	return int(math.Abs(float64(a)))
+func Abs[T constraints.Signed](a T) T {
```
Comment (apynes) [inline intutil/math.go:18]: missing test

File: intutil/math.go [hunk 5]
```diff
@@ -14,48 +15,60 @@ const (
...
+func Min[T constraints.Ordered](a, b T) T {
 	if a < b {
 		return a
 	}
 	return b
 }
 
 // Minv returns the smallest int passed in the variadic argument.
-func Minv(ints ...int) int {
+func Minv[T constraints.Ordered](ints ...T) T {
```
Comment (apynes) [inline intutil/math.go:35]: missing tests for this one as well

File: crypto/totp.go [hunk 6]
```diff
@@ -98,24 +100,47 @@ func TOTPCompare(key string, period time.Duration, adjust int, length int, chall
...
+
+	// add to vary if drift < 0
+	for i := drift; i < 0; i++ {
+		vary = append(vary, i-int(variance))
+	}
+
+	// sort vary to do 0, +1 && -1, +2 && -2, etc.
+	slices.SortFunc(vary, func(a, b int) bool {
+		return intutil.Abs(a) < intutil.Abs(b)
+	})
```
[the remark this answers is not in the document]
Comment (apynes) [reply crypto/totp.go:134]: we need to center the variance on the drift, we may want this in a separate function that can be tested.

File: crypto/totp.go [hunk 7]
```diff
@@ -98,24 +100,47 @@ func TOTPCompare(key string, period time.Duration, adjust int, length int, chall
...
+		vary = append(vary, i+int(variance))
+	}
+
+	// add to vary if drift < 0
+	for i := drift; i < 0; i++ {
+		vary = append(vary, i-int(variance))
+	}
+
+	// sort vary to do 0, +1 && -1, +2 && -2, etc.
+	slices.SortFunc(vary, func(a, b int) bool {
```
Comment (apynes) [inline crypto/totp.go:132]: kind of annoying that they did not make the sort functions fluent, but oh well.

**Review by apynes — CHANGES_REQUESTED** [review]
Looks good, just need some changes around the frame constructions and this will be good to go

**Review by apynes — APPROVED** [review]
LGTM :+1:

**Pull request merged**